### DISCUSSION TITLE
Custom features for individual users

### DISCUSSION THREAD

**Discussion by gilesknap** [devcontainers/community#60] [Ideas]
I work in an organization where we have many repos each of which will have a devcontainer.

I would like our developers to be able to customize their personal experience inside these devcontainers without affecting the repository itself. In addition I would like the same customization they choose to be available to them inside all of the repos.

Essentially I'd like all of our devcontainers to provide hooks that allow individuals to add host mounts, installs and startup scripts etc. For example some of our developers like zsh and want to have access to ohmyzsh files inside the devcontainer. Another requirement is to have an eternal history file that works inside/outside devcontainers.

Is this something that others are interested in? Does anyone have a solution to this?

Approaches I have attempted:
- mount the home directory from the host to `/hosthome` and execute `/hosthome/.devcontainer-rc/install.sh` which can soft link / copy / apt-get install as required
- use a local feature definition that is in `$/HOME/.devcontainer-rc/custom_feature/` and in `devcontainer.json` use `initializeCommand` to rsync it into the `devcontainer` folder as a first step. This feature can mount whatever parts of your home directory etc that you like and do anything else in install.sh

The first option is not great because it always mounts your whole home directory.
The second is a little convoluted and requires some mitigation for those that don't have a `.devcontainer-rc` folder (fetching an empty feature definition)

I'd be interested in any ideas on how to improve upon these.

**Comment by bamurtaugh** [comment]
Thanks for sharing this @gilesknap! I'd recommend sharing this in our dev container community Slack channel as well, as there's a broad set of folks from the dev container community who may have thoughts on this: https://aka.ms/dev-container-community.

**Comment by jpillor-macquarie** [comment]
Hey @gilesknap see https://code.visualstudio.com/docs/devcontainers/containers#_personalizing-with-dotfile-repositories to use dotfiles with your devcontainers

**Reply by gilesknap** [reply]
Thanks @jpillor-macquarie I was not aware of this feature.
I'll give this a try.

**Reply by evilhamsterman** [reply]
It works great combined with the default features and default extensions in VSCode. For instance I prefer fish and have a highly customized dotfiles I use with chezmoi. So I have VSCode set to install the fish feature pull in my dotfiles and run install.sh which installs chezmoi and applies it. I get my normal CLI environment wherever I go. Add the default extensions I get my favorite extensions too.

```json
{  
  "dev.containers.defaultFeatures": {  
    "ghcr.io/meaningful-ooo/devcontainer-features/fish:1": {}
  },
  "dev.containers.defaultExtensions": [
    "albert.TabOut"
  ],
}
```

**Comment by gilesknap** [comment]
It looks like the dotfile repo user setting covers most of what I need.

However, it is exclusively for configuration. What solutions are there for when we want to save state back to the host?
Two example uses are

- eternal shell history file that covers history across all devcontainers (and the host shells)
- ssh configuration, including updating known_hosts

**Reply by evilhamsterman** [reply]
There's various persistence features that save stuff to docker volumes that you could adapt. There's another request for optional mounts and such
